[FSDP] Option to summon on rank 0 only

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #73904
* __->__ #73903

Add an option to summon full parameter on rank0_only. writeback=True is not supported. 

Differential Revision: [D34706449](https://our.internmc.facebook.com/intern/diff/D34706449/)

rohan-varma committed:
[FSDP] Option to summon on rank 0 only

Add an option to summon full parameter on rank0_only. Note that as documented, when using this flag, writes are only guaranteed to persist for rank 0.

Differential Revision: [D34706449](https://our.internmc.facebook.com/intern/diff/D34706449/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Option to summon on rank 0 only"


Add an option to summon full parameter on rank0_only. Note that as documented, when using this flag, writes are only guaranteed to persist for rank 0.

Differential Revision: [D34706449](https://our.internmc.facebook.com/intern/diff/D34706449/)

[ghstack-poisoned]

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1147,15 +1150,26 @@ def summon_full_params(
+                materialized on only global rank 0. This means that within the
+                context, only rank 0 will have full parameters and the other
+                ranks will have sharded parameters. This further means that for
+                non-zero ranks, the `writeback` argument is ignored and
```
The comment is not complete?

fegin replied:
Instead of implicitly ignoring writeback, should we ask users to explicitly writeback to False?

fegin reviewed (commented):
LGTM. Just one comment about how the combination of the arguments summon_full_param() are handled.

zhaojuanmao reviewed (approved):
awesome, looks god!

rohan-varma committed:
Update on "[FSDP] Option to summon on rank 0 only"


Add an option to summon full parameter on rank0_only. Note that as documented, when using this flag, writes are only guaranteed to persist for rank 0.

Differential Revision: [D34706449](https://our.internmc.facebook.com/intern/diff/D34706449/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Option to summon on rank 0 only"


Add an option to summon full parameter on rank0_only. writeback=True is not supported. 

Differential Revision: [D34706449](https://our.internmc.facebook.com/intern/diff/D34706449/)

[ghstack-poisoned]
